qolmat/analysis/holes_characterization.py

```diff
@@ -88,7 +101,545 @@ def test(self, df: pd.DataFrame) -> float:
             diff_means = obs_mean - ml_means[list(tup_pattern)]
             inv_sigma_pattern = np.linalg.inv(ml_cov[:, tup_pattern][tup_pattern, :])
 
-            d0 += n_rows_pattern * np.dot(np.dot(diff_means, inv_sigma_pattern), diff_means.T)
+            d0 += n_rows_pattern * np.dot(
+                np.dot(diff_means, inv_sigma_pattern), diff_means.T
+            )
             degree_f += tup_pattern.count(True)
 
         return 1 - float(chi2.cdf(d0, degree_f))
+
+
+class PKLMTest(McarTest):
+    """
+    This class implements the PKLM test, a fully non-parametric, easy-to-use, and powerful test
+    for the missing completely at random (MCAR) assumption on the missingness mechanism of a
+    dataset. The null hypothesis is "The missing data mechanism is MCAR".
+
+    This test is applicable to mixed data (quantitative and categoricals) types.
+
+
+    If you're familiar with the paper, this implementation of the PKLM test was made for the
+    parameter size.resp.set=2 only.
+
+    References
+    ----------
+    Spohn, M. L., Näf, J., Michel, L., & Meinshausen, N. (2021). PKLM: A flexible MCAR test using
+    Classification. arXiv preprint arXiv:2109.10150.
+
+    Parameters
+    -----------
+    nb_projections : int
+        Number of projections.
+    nb_permutation : int
+        Number of permutations.
+    nb_trees_per_proj : int
+        Number of trees per projection.
+    compute_partial_p_values : bool
+        If true, compute the partial p-values.
+    exact_p_value : bool
+        If True, compute exact p-value.
+    encoder : OneHotEncoder or None, default=None
+        Encoder to convert non numeric pandas dataframe values to numeric values.
+    random_state : int, RandomState instance or None, default=None
+        Controls the randomness.
+        Pass an int for reproducible output across multiple function calls.
+    """
+
+    def __init__(
+        self,
+        nb_projections: int = 100,
+        nb_projections_threshold: int = 200,
+        nb_permutation: int = 30,
+        nb_trees_per_proj: int = 200,
+        compute_partial_p_values: bool = False,
+        exact_p_value: bool = False,
+        encoder: Union[None, OneHotEncoder] = None,
+        random_state: Union[None, int, np.random.RandomState] = None,
+    ):
+        super().__init__(random_state=random_state)
+        self.nb_projections = nb_projections
+        self.nb_projections_threshold = nb_projections_threshold
+        self.nb_permutation = nb_permutation
+        self.nb_trees_per_proj = nb_trees_per_proj
+        self.compute_partial_p_values = compute_partial_p_values
+        self.exact_p_value = exact_p_value
+        self.encoder = encoder
+
+        if self.exact_p_value:
+            self.process_permutation = self._parallel_process_permutation_exact
+        self.process_permutation = self._parallel_process_permutation
```

Comment:
Not sure it does what you want. 
I think you should move that line before the if statement, or in a else statement.